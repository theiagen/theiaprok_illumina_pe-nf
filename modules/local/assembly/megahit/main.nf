process MEGAHIT {
    tag "$meta.id"
    label "process_low"
    
    // This docker container works with nextflow, ours doesn't for some reason, but works in the WDL implementaion
    // Not sure what's going on, but this is a workaround for now
    container "community.wave.seqera.io/library/megahit_pigz:87a590163e594224"

    input:
    tuple val(meta), path(reads)
    val kmers

    output:
    tuple val(meta), path('*_megahit_contigs.fa'), emit: assembly_fasta
    tuple val(meta), path("*.log")        , emit: megahit_log
    path "versions.yml"                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def min_contig_len = params.min_contig_length ?: 1
    def memory_fraction = 0.9
    def k_list = kmers ? "--k-list ${kmers}" : ""
    def reads_input = meta.single_end ? "-r ${reads}" : "-1 ${reads[0]} -2 ${reads[1]}"
    
    """
    megahit \\
        ${reads_input} \\
        --min-contig-len ${min_contig_len} \\
        -m ${memory_fraction} \\
        -t ${task.cpus} \\
        -o megahit \\
        ${k_list} \\
        ${args}

    mv megahit/final.contigs.fa ${prefix}_megahit_contigs.fa
    mv megahit/log ${prefix}_megahit.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        megahit: \$(megahit --version 2>&1 | sed 's/^.*MEGAHIT v//' | sed 's/ .*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_megahit_contigs.fa
    touch ${prefix}_megahit.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        megahit: \$(megahit --version 2>&1 | sed 's/^.*MEGAHIT v//' | sed 's/ .*\$//')
    END_VERSIONS
    """
}