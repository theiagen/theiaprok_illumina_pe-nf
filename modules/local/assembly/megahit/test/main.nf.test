nextflow_process {
    name "Test Process MEGAHIT"
    script "../main.nf"
    process "MEGAHIT"
    
    test("megahit - test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = ""
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.assembly_fasta,
                    process.out.megahit_log,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("megahit - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = ""
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}