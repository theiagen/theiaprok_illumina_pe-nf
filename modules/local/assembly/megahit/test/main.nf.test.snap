{
    "megahit - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_megahit_contigs.fa:md5,8be3bc67b657be1fcc54248b9307c0fe"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_megahit.log:md5,ee8ecbfb5a7185815732f6cad41feaa0"
                ]
            ],
            [
                "versions.yml:md5,e3c0731297c9abe2f495ab6d541ac0e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T15:44:51.641090705"
    },
    "megahit - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_megahit_contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_megahit.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,e3c0731297c9abe2f495ab6d541ac0e6"
                ],
                "assembly_fasta": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_megahit_contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "megahit_log": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_megahit.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e3c0731297c9abe2f495ab6d541ac0e6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T15:45:03.715292871"
    }
}